### PR TITLE
Open with browser

### PR DESCRIPTION
**Pull request by superiorlu**

### CODE REVIEW

**Comment by superiorlu** [comment]
hi, i add a new feature: the report will be open with browser after generate report.

File: lib/rubycritic/browser.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
+require "rubycritic/configuration"
+require "rubycritic/platforms/linux"
+require "rubycritic/platforms/darwin"
+
+module Rubycritic
+  class Browser
+    attr_reader :browser_name
```
Comment (troessner) [inline lib/rubycritic/browser.rb]: Why not just "name"? The "browser_" seems unnecessary since it is scoped to browser already.

File: lib/rubycritic/browser.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    end
+
+    def platform
+      [Rubycritic::Platforms::Linux, Rubycritic::Platforms::Darwin].find(&:can_open?)
+    end
+
+    def open(report_path)
+      platform && platform.new(browser_name, report_path).open
+    end
+
```
Comment (troessner) [inline lib/rubycritic/browser.rb]: No newline here pls.

File: lib/rubycritic/commands/default.rb [hunk 3]
```diff
@@ -12,7 +13,8 @@ def initialize(paths)
       end
 
       def execute
-        report(critique)
+        return report(critique) if Config.format == :json
```
Comment (troessner) [inline lib/rubycritic/commands/default.rb]: This will fail soon since a PR with a consolereport ist almost done.
Also I feel a 

```Ruby
case Config.format
when ....
end
```

is more expressive since it makes it clear that we are choosing paths based on format whereas your "return" makes it seem like json would be a special case.

File: lib/rubycritic/commands/default.rb [hunk 4]
```diff
@@ -20,6 +22,15 @@ def critique
         RevisionComparator.new(@paths).set_statuses(analysed_modules)
       end
 
+      def open(report_path = nil)
```
Comment (troessner) [inline lib/rubycritic/commands/default.rb]: ControlArgumentSmell. Lets move this whole logic to the caller since he already knows what to do.

File: lib/rubycritic/generators/html_report.rb [hunk 5]
```diff
@@ -18,6 +18,7 @@ def generate_report
         create_directories_and_files
         copy_assets_to_report_directory
         puts "New critique at #{report_location}"
+        report_location
```
Comment (troessner) [inline lib/rubycritic/generators/html_report.rb]: Why this change?

File: lib/rubycritic/platforms/base.rb [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+      def self.can_open?
+        false
+      end
+
+      def browser
+        raise "not implement."
+      end
+
+      def open
+        raise "not implement."
```
Comment (troessner) [inline lib/rubycritic/platforms/base.rb]: You can replace this with a proper ruby exception:

```Ruby
raise NotImplementedError
```

File: lib/rubycritic/platforms/base.rb [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+  module Platforms
+    class Base
+      attr_reader :browser_name, :report_path
+
+      def initialize(browser_name, report_path)
+        @browser_name = browser_name
+        @report_path  = report_path
+      end
+
+      def self.can_open?
```
Comment (troessner) [inline lib/rubycritic/platforms/base.rb]: Why is this a class method? Feels awkward...

Comment (troessner) [inline lib/rubycritic/platforms/base.rb]: And shouldnt we raise here as well?

File: lib/rubycritic/browser.rb (hunk 1 above)
Comment (superiorlu) [inline lib/rubycritic/browser.rb]: yes, i think this no need with 'browser_', i will modify it.

File: lib/rubycritic/platforms/darwin.rb [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+module Rubycritic
+  module Platforms
+    class Darwin < Rubycritic::Platforms::Base
+      def self.can_open?
+        platform =~ /darwin/
+      end
+
+      def application
+        if browser_name =~ /chrome/i
```
Comment (troessner) [inline lib/rubycritic/platforms/darwin.rb]: We are mixing code and data here - how about just using a hash here and get rid of those ugly "if"s?

File: lib/rubycritic/platforms/linux.rb [hunk 9]
```diff
@@ -0,0 +1,29 @@
+require "rubycritic/platforms/base"
+
+module Rubycritic
+  module Platforms
+    class Linux < Rubycritic::Platforms::Base
```
Comment (troessner) [inline lib/rubycritic/platforms/linux.rb]: You can drop the RubyCritic::Platforms part here and above.

File: lib/rubycritic/generators/html_report.rb (hunk 5 above)
Comment (superiorlu) [inline lib/rubycritic/generators/html_report.rb]: because it want to get report path before report will be open in Command::Default class , so there is another way?

File: lib/rubycritic/commands/default.rb (hunk 3 above)
Comment (superiorlu) [inline lib/rubycritic/commands/default.rb]: good suggestion.

File: lib/rubycritic/platforms/linux.rb (hunk 9 above)
Comment (superiorlu) [inline lib/rubycritic/platforms/linux.rb]: okay, i will drop it.

File: lib/rubycritic/cli/options.rb [hunk 10]
```diff
@@ -39,6 +39,10 @@ def parse
             @suppress_ratings = true
           end
 
+          opts.on("-O ", "--open", "Open report with browser") do |open_with|
```
Comment (troessner) [inline lib/rubycritic/cli/options.rb]: Let's add all the possibilities here a user can enter, otherwise it's just happy guessing ;)

File: lib/rubycritic/commands/default.rb [hunk 11]
```diff
@@ -12,7 +13,12 @@ def initialize(paths)
       end
 
       def execute
-        report(critique)
+        case Config.format
+        when :json
+          report(critique)
+        else
+          open(report(critique))
+        end
```
Comment (troessner) [inline lib/rubycritic/commands/default.rb]: Yes, nice!

Comment (troessner) [inline lib/rubycritic/commands/default.rb]: Since [this one](https://github.com/whitesmith/rubycritic/pull/91) is close to merge, let's update this to:

```Ruby
when :json
  report(critique)
when :html
  open(report(critique))
else
  raise "Unknown output format #{Config.format}"
end
```

File: lib/rubycritic/platforms/base.rb [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+      def browser
+        raise NotImplementedError.new("The #{self.class} class must implement the #{__method__} method.")
+      end
+
+      def open(_)
+        raise NotImplementedError.new("The #{self.class} class must implement the #{__method__} method.")
+      end
+
+      def application
+        raise NotImplementedError.new("The #{self.class} class must implement the #{__method__} method.")
```
Comment (troessner) [inline lib/rubycritic/platforms/base.rb]: Those error messages are very repetitive - given that "normal" user never see this message I'd rather go for just

```Ruby
raise NotImplementedError
```
with no explicit error message to cut down repetition and improve readability.

File: lib/rubycritic/generators/html_report.rb (hunk 5 above)
Comment (troessner) [inline lib/rubycritic/generators/html_report.rb]: Ah, yeah, I missed that, nevermind ;)

File: lib/rubycritic/commands/default.rb [hunk 13]
```diff
@@ -20,6 +26,15 @@ def critique
         RevisionComparator.new(@paths).set_statuses(analysed_modules)
       end
 
+      def open(report_path)
```
Comment (troessner) [inline lib/rubycritic/commands/default.rb]: This method will only be called when we use a html_formatter, right?
This means there will **always** be a report_path, won't it?
Cause then we can drop the argument check and delegate this call directly to `browser`.

File: lib/rubycritic/commands/default.rb (hunk 11 above)
Comment (troessner) [inline lib/rubycritic/commands/default.rb]: I just merged the PR mentioned above, please rebase against master and then lets add handling :console_report here.

Comment (superiorlu) [inline lib/rubycritic/commands/default.rb]: nice.

**Comment by superiorlu** [comment]
hi troessner. 
i find a question： when i run ```rubycritic -p new_path ```， it seemed like run twice.
qq20160117-0 2x

**Comment by troessner** [comment]
Yes, that's a [known problem](https://github.com/whitesmith/rubycritic/issues/94).
So it didn't originate in your PR ;)
Meaning: Just ignore that and let's fix this separately.

**Comment by superiorlu** [comment]
okay. we will fix it.

**Comment by superiorlu** [comment]
on 15th Jan , i published a topic about ```rubycritic``` on``` ruby china``` forum.
 [https://ruby-china.org/topics/28746](https://ruby-china.org/topics/28746). 
it hope that more pepole can use ```rubycritic``` analysis ruby code.

**Comment by troessner** [comment]
Ok, this looks very good now, love it!

Now to the last steps, we need to 

[1] rebase against master since we have merge conflicts 
[2] squash to one commit
[3] final review

[1] and [2] are up to you, I'd tackle 3 then.
Please write a comment here when you're finished with [1] and [2] since github doesnt send out notifications for those events.

**Comment by superiorlu** [comment]
Thanks for your help @troessner .
i will rebase and squash the commit.

**Comment by superiorlu** [comment]
hi @troessner, i  have done it :)

File: lib/rubycritic/cli/options.rb [hunk 14]
```diff
@@ -40,6 +40,16 @@ def parse
             @suppress_ratings = true
           end
 
+          opts.on("-O ", "--open", [:'chrome', :firefox, :safari, :chromium],
```
Comment (troessner) [inline lib/rubycritic/cli/options.rb]: 2 points:

1.) The quotation marks for

>> :'chrome'

are unnecessary and seem to be a remnant.

2.) Can we make the list of supported browsers a constant so that we can reference that here and [here](https://github.com/whitesmith/rubycritic/pull/100/files#diff-0710b2dc2bc4fa0cf5b4aa9a98870c48R12)

File: lib/rubycritic/commands/default.rb [hunk 15]
```diff
@@ -12,7 +13,15 @@ def initialize(paths)
...
       def execute
-        report(critique)
+        case Config.format
+        when :json
+          report(critique)
+        when :html
+          report(critique)
+          open
+        else
+          raise "Unknown output format #{Config.format}"
```
Comment (troessner) [inline lib/rubycritic/commands/default.rb]: This will now crash on our newly merged "ConsoleReport", won't it?
Did you rebase against master?

File: lib/rubycritic/platforms/base.rb [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+      end
+
+      def browser_hash
+        raise NotImplementedError
+      end
+
+      def platform
+        @platform ||= begin
+          if RUBY_PLATFORM =~ /java/
+            `uname -s`.downcase
```
Comment (troessner) [inline lib/rubycritic/platforms/base.rb]: Wouldn't this crash for somebody running JRuby on windows?
How about doing something like [here](http://stackoverflow.com/questions/11784109/detecting-operating-systems-in-ruby)?

File: lib/rubycritic/platforms/darwin.rb [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+module Rubycritic
+  module Platforms
+    class Darwin < Base
+      def can_open?
+        platform =~ /darwin/
+      end
+
+      def browser_hash
+        @browser_hash ||= begin
+          hash = { :chrome => "Google Chrome", :firefox => "firefox", :safari => "safari" }
```
Comment (troessner) [inline lib/rubycritic/platforms/darwin.rb]: Should be a constant, see my other comments above.

File: lib/rubycritic/platforms/darwin.rb [hunk 18]
```diff
@@ -0,0 +1,29 @@
+require "rubycritic/platforms/base"
+
+module Rubycritic
+  module Platforms
+    class Darwin < Base
+      def can_open?
+        platform =~ /darwin/
+      end
+
+      def browser_hash
```
Comment (troessner) [inline lib/rubycritic/platforms/darwin.rb]: No hungarian notation please. How about "browser_map"? ("map" is some kind of generic interface while "hash" is a concrete type).

File: lib/rubycritic/platforms/darwin.rb [hunk 19]
```diff
@@ -0,0 +1,29 @@
...
+        @browser_hash ||= begin
+          hash = { :chrome => "Google Chrome", :firefox => "firefox", :safari => "safari" }
+          hash.default = "safari"
+          hash
+        end
+      end
+
+      def application
+        browser_hash[browser_name.to_sym]
```
Comment (troessner) [inline lib/rubycritic/platforms/darwin.rb]: The "to_sym" seems like an unnecessary method call, isn't it?

File: lib/rubycritic/platforms/darwin.rb [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+
+module Rubycritic
+  module Platforms
+    class Darwin < Base
+      def can_open?
+        platform =~ /darwin/
+      end
+
+      def browser_hash
+        @browser_hash ||= begin
```
Comment (troessner) [inline lib/rubycritic/platforms/darwin.rb]: If you use a constant like suggested above:

```Ruby
BROWSER_MAP = {}.tap do |hash|
  hash = { :chrome => "Google Chrome", :firefox => "firefox", :safari => "safari" }
  hash.default = "safari"
end

**Comment by troessner** [comment]
@superiorlu couple of more comments I'm afraid, but we're almost good to merge this one, let's stay the course!

File: lib/rubycritic/platforms/darwin.rb (hunk 20 above)
Comment (superiorlu) [inline lib/rubycritic/platforms/darwin.rb]: :+1:

**Comment by superiorlu** [comment]
@troessner  okay, i wil do that with your suggestions. Thanks for your patience.

**Comment by superiorlu** [comment]
@troessner i have refactor the code and rebase the commit.
Thanks for your help.

File: test/lib/rubycritic/browser_test.rb [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+require "rubycritic/platforms/darwin"
+require "rubycritic/generators/html_report"
+
+describe Rubycritic::Browser do
+  before do
+    @report_path = "tmp/rubycritic/overview.html"
+    @browser = Rubycritic::Browser.new @report_path
+  end
+
+  describe "open report on diffrent platform" do
```
Comment (Onumis) [inline test/lib/rubycritic/browser_test.rb]: typo: `different`

File: lib/rubycritic/platforms/darwin.rb [hunk 22]
```diff
@@ -0,0 +1,26 @@
+require "rubycritic/platforms/base"
+
+module Rubycritic
+  module Platforms
+    class Darwin < Base
+      BROWSER_MAP = {}.tap do |hash|
+        hash.merge! :chrome => "Google Chrome", :firefox => "firefox", :safari => "safari"
```
Comment (Onumis) [inline lib/rubycritic/platforms/darwin.rb]: just to be sure, is it really `Google Chrome` instead of `google-chrome`? (I don't have a mac so, I can't assess this myself)

File: test/lib/rubycritic/browser_test.rb [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+        @linux.browser.must_equal "google-chrome"
+      end
+
+      it "open with chromium browser " do
+        Rubycritic::Config.stubs(:open_with).returns(:chromium)
+        Rubycritic::Browser.any_instance.stubs(:platform).returns(Rubycritic::Platforms::Linux.new)
+        @linux = @browser.platform
+        @linux.browser.must_equal "chromium-browser"
+      end
+    end
```
Comment (Onumis) [inline test/lib/rubycritic/browser_test.rb]: I think we can apply some DRY here:

```
let(:report_path) { "tmp/rubycritic/overview.html" }
let(:browser) { Rubycritic::Browser.new report_path }

before do
  Rubycritic::Browser.any_instance.stubs(:platform).returns(target_platform)
end

describe "open report on different platform" do
  context "on linux" do
    let(:target_platform) { Rubycritic::Platforms::Linux.new }
    it "open with default browser" do
      Rubycritic::Config.set
      platform.browser.must_equal "firefox"
    end
    #...  
  end
  context "on darwin" do
    let(:target_platform) { Rubycritic::Platforms::Linux.new }
    #...
  end
end
```

I haven't tested this, but it should work

File: features/command_line_interface/options.feature [hunk 24]
```diff
@@ -29,6 +29,7 @@ Feature: Rubycritic can be controlled using command-line options
           -m, --mode-ci                    Use CI mode (faster, but only analyses last commit)
               --deduplicate-symlinks       De-duplicate symlinks based on their final target
               --suppress-ratings           Suppress letter ratings
+          -O, --open                       Open html report with browser: [:chrome, :chromium, :firefox, :safari]
```
Comment (Onumis) [inline features/command_line_interface/options.feature]: Maybe we should say here `(Leave blank to open system default)`?

File: lib/rubycritic/platforms/darwin.rb (hunk 22 above)
Comment (superiorlu) [inline lib/rubycritic/platforms/darwin.rb]: @Onumis  yes, i test it in my mac osx 10.10.5.
 [http://superuser.com/questions/350309/how-can-i-open-a-url-in-google-chrome-from-the-terminal-in-os-x](http://superuser.com/questions/350309/how-can-i-open-a-url-in-google-chrome-from-the-terminal-in-os-x)

File: test/lib/rubycritic/browser_test.rb (hunk 23 above)
Comment (superiorlu) [inline test/lib/rubycritic/browser_test.rb]: nice. that looks better :)

File: features/command_line_interface/options.feature (hunk 24 above)
Comment (superiorlu) [inline features/command_line_interface/options.feature]: on linux, firefox is default. 
on mac osx, safari is default.
so i think default may be not important. just run ```rubycritic``` :smile:

File: test/lib/rubycritic/browser_test.rb (hunk 23 above)
Comment (Onumis) [inline test/lib/rubycritic/browser_test.rb]: Sure, push those changes and fix the typo on "diffrent" and it'll be good to go :)

Comment (superiorlu) [inline test/lib/rubycritic/browser_test.rb]: Okay, thanks. @Onumis

**Comment by superiorlu** [comment]
Hi @Onumis i have fixed it and rebase commit. Thanks.

File: lib/rubycritic/platforms/base.rb [hunk 25]
```diff
@@ -0,0 +1,47 @@
...
+          case host_os
+          when /mswin|msys|mingw|cygwin|bccwin|wince|emc/
+            :windows
+          when /darwin|mac os/
+            :darwin
+          when /linux/
+            :linux
+          when /solaris|bsd/
+            :unix
+          end
```
Comment (troessner) [inline lib/rubycritic/platforms/base.rb]: Yes, that is a LOT nicer now :)

File: lib/rubycritic/platforms/base.rb [hunk 26]
```diff
@@ -0,0 +1,47 @@
...
+        end
+      end
+
```
Comment (troessner) [inline lib/rubycritic/platforms/base.rb]: This will break the build after rebase. Now newline here pls.

File: lib/rubycritic/platforms/darwin.rb [hunk 27]
```diff
@@ -0,0 +1,26 @@
...
+      end
+
+      def can_open?
+        platform == :darwin
+      end
+
+      def application
+        BROWSER_MAP[browser_name]
+      end
+      alias_method :browser, :application
```
Comment (troessner) [inline lib/rubycritic/platforms/darwin.rb]: This will also break the build after rebase because rubocop will complain about that, sorry ;)
Should be `alias`.

File: lib/rubycritic/platforms/darwin.rb [hunk 28]
```diff
@@ -0,0 +1,26 @@
...
+
+      def open(report_path)
+        system("open -a '#{application}' #{report_path}")
+      end
+    end
+
```
Comment (troessner) [inline lib/rubycritic/platforms/darwin.rb]: See above.

**Comment by troessner** [comment]
Actually you can scratch my comments about breaking the build since my PR is not merged yet :)

**Comment by troessner** [comment]
Merged! Great job, thanks!

**Comment by superiorlu** [comment]
Thanks @troessner @Onumis i have learned more than i expected. Thanks your help and patience.
This is my first project which contributed  to in github,  your advice makes me more confident. Thanks again. :)
